src/ocean/net/server/unix/CommandRegistry.d

```diff
@@ -0,0 +1,134 @@
+
+    Copyright:
+        Copyright (c) 2009-2018 Sociomantic Labs GmbH.
+        All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.net.server.unix.CommandRegistry;
+
+import ocean.transition;
+
+/// ditto
+public class CommandsRegistry
+{
+    import ocean.core.array.Transformation: filter, split;
+    import ocean.core.Buffer;
+    import ocean.core.Enforce;
+    import ocean.core.Verify;
+    import ocean.text.convert.Integer;
+
+    /// Alias for our interactive handler delegate.
+    public alias void delegate ( cstring[],
+            void delegate (cstring),
+            void delegate (ref mstring) ) InteractiveHandler;
+
+    /// Alias for our non-interactive handler delegate.
+    public alias void delegate ( cstring[],
+            void delegate (cstring) ) Handler;
+
+    /// Our registered map of interactive handlers by command.
+    private InteractiveHandler[istring] interactive_handlers;
+
+    /// Our registered map of handlers by command.
+    private Handler[istring] handlers;
+
+    /// Stores the command arguments split by " ";
+    private cstring[] args_buf;
+
+    /***************************************************************************
+
+        Receives all commands from the socket and splits the command by " ".
+        If a matching command is registered then the command will be called with
+        the remaining arguments. This method will be called by the
+        UnixSocketListener whenever a command is received.
+
+        Params:
+            command = The command received by the unix socket.
+            args = The arguments provided with the command.
+            send_response = Delegate to call with response string.
+            wait_reply = Delegate to call to obtain the reply from the user
+
+    ***************************************************************************/
+
+    public void handle ( cstring command, cstring args,
+                 void delegate (cstring) send_response,
+                 void delegate (ref mstring buf) wait_reply )
+    {
+        if (auto handler = command in this.interactive_handlers)
+        {
+            split(args, " ", this.args_buf);
+            scope predicate = (cstring v) { return v.length; };
+            auto arguments = filter(this.args_buf[], predicate, this.args_buf);
+            (*handler)(arguments, send_response, wait_reply);
+        }
+        else if (auto handler = command in this.handlers)
+        {
+            split(args, " ", this.args_buf);
+            scope predicate = (cstring v) { return v.length; };
+            auto arguments = filter(this.args_buf[], predicate, this.args_buf);
```

Comment:
This will allocate on each unix socket command, surely we don't want that? https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/core/array/Transformation.d#L375

Same for filter/split calls few lines above

Reply:
Ah good catch! This is old code so I didn't pay attention.

Reply:
Very nice how this method actually receives a buffer but then ruins it all :-)